conda_smithy/lint_recipe.py

```diff
@@ -568,6 +569,45 @@ def lintify(meta, recipe_dir=None, conda_forge=False):
                 "upstream for the package's Python compatibility."
             )
 
+    # 26: pin_subpackage is for subpackages and pin_compatible is for
+    # non-subpackages of the recipe
+    subpackage_names = []
+    for out in outputs_section:
+        if "name" in out:
+            subpackage_names.append(out["name"])  # explicit
+    if "name" in package_section:
+        subpackage_names.append(package_section["name"])  # implicit
+
+    def check_pins(pinning_section):
+        if pinning_section is None:
+            return
+        for pin in fnmatch.filter(pinning_section, "compatible_pin*"):
+            if pin.split()[-1] in subpackage_names:
+                lints.append(
```

Comment:
Can we make this a `hint` instead of a `lint`? Some package names have `jinja` and the simple parser in `lint_recipe` sometimes gets it wrong.

Reply:
You are saying that something like this 

```yaml
outputs:
  - name: lib{{ apackagename }}
```

could cause problems?

Reply:
Depends on whether `apackagename` is defined directly or using other conda-build jinja variables.

Reply:
My understanding is that hints are non-breaking errors (spellcheck, license ID, using noarch), and lints are for breaking errors (invalid names, incorrect version specifiers, missing maintainers).

I think that incorrect usage of pinning expressions counts as a breaking error because the pin will not work as intended, but conda-build/jinja won't say anything about it.

What if I add the list of known sub-packages to the lint message so that it's easier to debug/ignore if the jinja templating doesn't work correctly? The message would be something like this:

> pin_subpackage should be used instead of pin_compatible for `apackage` because it is one of the known outputs of this recipe: ['package-split', 'apackage', 'libapackage', 'apackage-python']

Reply:
Yes, that and as long as you commit to fix false positives.

Reply:
😆 Am I allowed to demote this rule from lint to hint later if I discover something I can't fix?

Reply:
Sure. :smiley:

Reply:
Can you check that https://github.com/conda-forge/ctng-compilers-feedstock/blob/main/recipe/meta.yaml works? I'm pretty sure this PR would add a lint

Reply:
No lint detected.

Reply:
This is the list of detected subpackages:

['libgcc-devel_', 'libstdcxx-devel_', 'gcc_impl_', 'conda-gcc-specs', 'gxx_impl_', 'gfortran_impl_', 'libstdcxx-ng', 'libsanitizer', 'libgomp', 'libgcc-ng', '_openmp_mutex', 'libgfortran', 'libgfortran-ng', 'gcc_compilers']

This means some of the jinja templates were blank or empty.

Reply:
Ah, I guess since the package name and the `pin_subpackage` were both missing the plarform, no lints detected.

Reply:
jinja variables defined externally (from conda_build_config.yaml) are replaced with empty_string because they depend on the variant.